nearcore/src/runtime/mod.rs

```diff
@@ -712,23 +713,39 @@ impl RuntimeAdapter for NightshadeRuntime {
         gas_limit: Gas,
         epoch_id: &EpochId,
         shard_id: ShardId,
-        state_root: StateRoot,
+        storage_config: RuntimeStorageConfig,
         next_block_height: BlockHeight,
-        pool_iterator: &mut dyn PoolIterator,
+        transaction_groups: &mut dyn TransactionGroupIterator,
         chain_validate: &mut dyn FnMut(&SignedTransaction) -> bool,
         current_protocol_version: ProtocolVersion,
         time_limit: Option<Duration>,
     ) -> Result<PreparedTransactions, Error> {
         let start_time = std::time::Instant::now();
         let shard_uid = self.get_shard_uid_from_epoch_id(shard_id, epoch_id)?;
-        let mut state_update = self.tries.new_trie_update(shard_uid, state_root);
+
+        let mut trie = match storage_config.source {
+            StorageDataSource::Recorded(storage) => Trie::from_recorded_storage(
+                storage,
+                storage_config.state_root,
+                storage_config.use_flat_storage,
+            ),
+            _ => self.tries.get_trie_for_shard(shard_uid, storage_config.state_root),
```

Comment:
using "catch all" pattern could be error-prone here since `DbTrieOnly` is not actually supported, but will be treated as `Db` when set
I suggest explicitly handing every case with panic on `DbTrieOnly`